dialogue-core/src/main/java/com/palantir/dialogue/core/RetryOtherValidatingChannel.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.RateLimiter;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.UnsafeArg;
+import java.util.List;
+import java.util.Optional;
+import javax.ws.rs.core.HttpHeaders;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class RetryOtherValidatingChannel implements Channel {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryOtherValidatingChannel.class);
+    private static final RateLimiter UNKNOWN_RETRY_OTHER_URI = RateLimiter.create(1);
+
+    private final Channel delegate;
+    private final List<String> uris;
+    private final UnsafeArg<List<String>> unsafeArg;
+
+    private RetryOtherValidatingChannel(Channel delegate, List<String> uris) {
+        this.delegate = delegate;
+        this.uris = uris;
+        this.unsafeArg = UnsafeArg.of("uris", uris);
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return DialogueFutures.transform(delegate.execute(endpoint, request), response -> {
+            Optional<String> maybeRetryOtherHost = response.getFirstHeader(HttpHeaders.LOCATION);
+            if (maybeRetryOtherHost.isPresent()) {
+                String retryOtherHost = maybeRetryOtherHost.get();
+                if (!isKnown(maybeRetryOtherHost.get()) && UNKNOWN_RETRY_OTHER_URI.tryAcquire()) {
+                    log.info("Unknown Location header", UnsafeArg.of("location", retryOtherHost), unsafeArg);
+                }
+            }
+
+            return response;
+        });
+    }
+
+    private boolean isKnown(String retryOtherUri) {
+        for (int i = 0; i < uris.size(); i++) {
+            if (uris.get(i).equals(retryOtherUri)) {
```

Comment:
Is the retry-other uri always equivalent to the configured URI? I suppose we'll know if this fires.

Mostly curious if the values may include additional path segments, or potentially missing the `/api` prefix. URI parsing can be a little bit tricky as the same value may be encoded a couple different ways. Perhaps we should extract the `host` component of the URI and compare those?

Reply:
I wanted to dig up what conjure-okhttp used to do, but I'm pretty sure it would've just used the same method as here.

Reply:
I started by trying to parse the urls, but decided this should probably be the contract IF we were to every implement this: it needs to match what we know about, otherwise it's a little wonky. Like as a strawman, I'm talking to a service using uris IT HAS published: it would make sense to me that IF that service sends me a retryOther, it should be the same uri format as it would've published them to me to use.

Reply:
Hmm maybe it was doing some more checks: https://github.com/palantir/conjure-java-runtime/blob/develop/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/UrlSelectorImpl.java

Reply:
I would still maintain that this is slightly weird API, they should be exchanging equivalent uris, so the client can just do a simple search.

Reply:
Given that this isn't connected to correctness or security, rather validation and debugging, I think the question we'd like to answer is "Is this URI definitely wrong?" as opposed to "Is this URI exactly right?". Not worth thinking about the subtle edge cases we don't need to.

Reply:
ok, in that case host is probably all we want